File: sdks/typescript/packages/native/src/public-key-assertion.ts
/** Helpers for the Native public-key direct-issue assertion. */

import type { DirectIssuePublicKeyRequest, PublicKeyCredential } from "./declare.js";

export const PUBLIC_KEY_ASSERTION_AUDIENCE = "sudomimus-native-public-key";
export const PUBLIC_KEY_ASSERTION_TYPE = "vnd.sudomimus.public-key-assertion+jwt";
export const PUBLIC_KEY_AUTHORIZATION_SCHEME = "SudomimusPublicKeyJWT";

const encoder = new TextEncoder();

const base64Url = (value: Uint8Array): string => {

    const base64 = globalThis.btoa(String.fromCharCode(...value));
    return base64.replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");
};

export type SignedPublicKeyRequest = {
    /** Exact UTF-8 request body covered by `requestHash`. */
    body: string;
    /** Complete Authorization header value. */
    authorization: string;
};

/**
 * Serialize a public-key request once, bind its exact bytes into a compact
 * EdDSA assertion, and return both values so callers cannot accidentally send
 * bytes different from those that were signed.
 */
export const signPublicKeyRequest = async (
    request: DirectIssuePublicKeyRequest,
    credential: PublicKeyCredential,
    now: Date = new Date(),
    jtiBytes?: Uint8Array,
): Promise<SignedPublicKeyRequest> => {

    const body = JSON.stringify(request);
    const bodyBytes = encoder.encode(body);
    const requestDigest = new Uint8Array(await globalThis.crypto.subtle.digest("SHA-256", bodyBytes));
    const nonce = jtiBytes ?? globalThis.crypto.getRandomValues(new Uint8Array(16));

    if (nonce.byteLength !== 16) {

        throw new TypeError("jtiBytes must contain exactly 16 random bytes.");
    }

    const issuedAt = Math.floor(now.getTime() / 1000);
    const header = {
        alg: "EdDSA",
        typ: PUBLIC_KEY_ASSERTION_TYPE,
        kid: credential.keyId,
    } as const;
    const claims = {
        iss: credential.keyId,
        aud: PUBLIC_KEY_ASSERTION_AUDIENCE,
        iat: issuedAt,
        exp: issuedAt + 60,
        jti: base64Url(nonce),
        requestHash: base64Url(requestDigest),
    } as const;
    const headerSegment = base64Url(encoder.encode(JSON.stringify(header)));
    const claimsSegment = base64Url(encoder.encode(JSON.stringify(claims)));
    const signingInput = encoder.encode(`${headerSegment}.${claimsSegment}`);
    const signature = await credential.sign(signingInput);

    if (signature.byteLength !== 64) {

        throw new TypeError("Ed25519 signatures must contain exactly 64 bytes.");
    }

    const assertion = `${headerSegment}.${claimsSegment}.${base64Url(signature)}`;
    return {
        body,
        authorization: `${PUBLIC_KEY_AUTHORIZATION_SCHEME} ${assertion}`,
    };
};

File: sdks/python/packages/sudomimus-native/src/sudomimus_native/public_key_assertion.py
"""Signed assertion helpers for Native public-key direct issue."""

from __future__ import annotations

import base64
import hashlib
import json
import secrets
import time
from collections.abc import Callable
from dataclasses import dataclass

from ._generated.models import DirectIssuePublicKeyRequest

PUBLIC_KEY_ASSERTION_AUDIENCE = "sudomimus-native-public-key"
PUBLIC_KEY_ASSERTION_TYPE = "vnd.sudomimus.public-key-assertion+jwt"
PUBLIC_KEY_AUTHORIZATION_SCHEME = "SudomimusPublicKeyJWT"


def _base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


@dataclass(frozen=True, slots=True)
class PublicKeyCredential:
    """Registered Ed25519 key ID and raw-signature callback."""

    key_id: str
    sign: Callable[[bytes], bytes]


@dataclass(frozen=True, slots=True)
class SignedPublicKeyRequest:
    """Exact request bytes and matching Authorization value."""

    body: bytes
    authorization: str


def sign_public_key_request(
    request: DirectIssuePublicKeyRequest,
    credential: PublicKeyCredential,
    *,
    now: int | None = None,
    jti_bytes: bytes | None = None,
) -> SignedPublicKeyRequest:
    """Serialize once and bind those exact bytes into a compact EdDSA JWT."""
    body = request.model_dump_json(exclude_none=True).encode("utf-8")
    issued_at = int(time.time()) if now is None else now
    nonce = secrets.token_bytes(16) if jti_bytes is None else jti_bytes
    if len(nonce) != 16:
        raise ValueError("jti_bytes must contain exactly 16 random bytes")

    header = {
        "alg": "EdDSA",
        "typ": PUBLIC_KEY_ASSERTION_TYPE,
        "kid": credential.key_id,
    }
    claims = {
        "iss": credential.key_id,
        "aud": PUBLIC_KEY_ASSERTION_AUDIENCE,
        "iat": issued_at,
        "exp": issued_at + 60,
        "jti": _base64url(nonce),
        "requestHash": _base64url(hashlib.sha256(body).digest()),
    }
    encode_json = lambda value: json.dumps(  # noqa: E731
        value, separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")
    signing_input = (
        _base64url(encode_json(header)) + "." + _base64url(encode_json(claims))
    ).encode("ascii")
    signature = credential.sign(signing_input)
    if len(signature) != 64:
        raise ValueError("Ed25519 signatures must contain exactly 64 bytes")
    assertion = signing_input.decode("ascii") + "." + _base64url(signature)
    return SignedPublicKeyRequest(
        body=body,
        authorization=f"{PUBLIC_KEY_AUTHORIZATION_SCHEME} {assertion}",
    )
